benchmarks/performance-poetry/complex-poetry/src/main/java/com/squareup/benchmarks/performance/complex/poetry/PerformancePoetryComponent.kt

```diff
@@ -37,9 +45,121 @@ class PerformancePoetryComponent : ViewModel() {
+      }
+
+      sessionRenderStats[rootSessionId]!!.apply {
+        // Update stats for this render pass with this node.
+        val renderStateString = renderState.toString()
+        val lastState = nodeStates[session.sessionId]
+        if (lastState == null) {
+          nodesRenderedDifferent += 1
+        } else {
+          if (lastState.contentEquals(renderStateString)) {
+            nodesRenderedTheSame += 1
+          } else {
+            nodesRenderedDifferent += 1
+          }
+        }
+        nodeStates[session.sessionId] = renderStateString
+      }
+
+      return proceed(renderProps, renderState, null).also {
+        if (tracedId) {
+          Trace.endSection()
+        }
+
+        if (root) {
+          // Double end section for the 'total' and the one for this render pass.
+          Trace.endSection()
+          Trace.endSection()
+          // Print stats for this render pass.
+          val totalNodes =
+            sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent +
+              sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame
+          Timber.i("~~~~~~ Workflow Render Pass Stats Begin ~~~~~~")
+          Timber.i(
+            "Nodes rendered identically in this pass:" +
+              " ${sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame}"
+          )
+          Timber.i(
+            "Nodes rendered differently in this pass:" +
+              " ${sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent}"
+          )
+          Timber.i("Total Nodes: $totalNodes")
+          Timber.i(
+            "Fresh Rendering Ratio:" +
+              " ${
+              sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent.toDouble() /
+                totalNodes.toDouble()
+              }"
+          )
+          Timber.i("~~~~~~ Workflow Render Pass Stats End ~~~~~~")
+          totalFresh += sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent
+          totalStale += sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame
+          Timber.i("--------- Workflow Renderings Cumulative Stats ---------")
+          Timber.i("Total Render Passes So Far: $totalRenderPasses")
+          Timber.i(
+            "Total Fresh Rendering Ratio: ${
+            totalFresh.toDouble() / (totalFresh + totalStale).toDouble()
+            }"
+          )
+          totalRenderPasses += 1
+        }
+      }
+    }
+  }
+
   fun poetryModelFactory(owner: AppCompatActivity): PoetryModel.Factory =
     PoetryModel.Factory(
       owner,
-      loadingGatekeeperForPoems
+      loadingGatekeeperForPoems,
+      performanceTraceInterceptor
+    )
+
+  companion object {
+    // Update this value if render passes for the scenario under test change.
+    // Can do so with `adb logcat | grep PerformancePoetryComponent` and look through the
+    // Render Pass Stats.
+    const val TOTAL_RENDER_PASSES_OPEN_RAVEN_NAVIGATE = 32
```

Comment:
Should you throw or something if the list grows bigger than this, so we'll notice?

Reply:
And/or we could make this a much bigger number (1000 say), so that it won't matter so much if it climbs a bit.

I'm presuming you're trying to avoid a hit for expanding the array in mid-benchmark.

Reply:
no I think this has to do with the number of sections to look for in the final perfetto file, so you do want to know the exact number.

Alternatively you could also ditch macrobenchmark's provided results and query perfetto with the trace directly, which we then give you the list of all sections and wouldn't need to anticipate that. I wish this was possible directly within macrobenchmark but it's not yet, so you'd have to locate the trace file and do this after the fact.

Reply:
@pyricau is right that we should keep this exact for now.